src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/DeserializationHelpers.cs

```diff
@@ -57,6 +58,32 @@ internal static class DeserializationHelpers
             }
         }
 
+
+        internal static void ApplyInstanceAnnotations(object resource, IEdmStructuredTypeReference structuredType, ODataResource oDataResource,
+            ODataDeserializerProvider deserializerProvider, ODataDeserializerContext readContext)
+        {   
+            PropertyInfo propertyInfo = EdmLibHelpers.GetInstanceAnnotationsDictionary(structuredType.StructuredDefinition(), readContext.Model);
+            if (propertyInfo == null)
+            {
+                return;
+            }
+
+            IDelta delta = resource as IDelta;
+            object value;
+            if (delta != null)
+            {                
+                delta.TryGetPropertyValue(propertyInfo.Name, out value);
+            }
+            else
+            {
+                value = propertyInfo.GetValue(resource);
+            }
```

Comment:
delta and value are calculated from resource and propertyInfo and only used within GetAnnotationContainer.  So, rather than calculating them here and passing them in as individual arguments to GetAnnotationContainer, you should move this code into GetAnnotationContainer and just pass the resource and propertyInfo to that method.